crates/signer/src/signer.rs

```diff
@@ -0,0 +1,218 @@
+use crate::{Result, Signature};
+use alloy_primitives::{eip191_hash_message, Address, B256};
+use async_trait::async_trait;
+use auto_impl::auto_impl;
+
+#[cfg(feature = "eip712")]
+use alloy_sol_types::{Eip712Domain, SolStruct};
+
+/// Asynchronous Ethereum signer.
+///
+/// All provided implementations rely on [`sign_hash`](Signer::sign_hash). A signer may not always
+/// be able to implement this method, in which case it should return
+/// [`UnsupportedOperation`](crate::Error::UnsupportedOperation), and implement all the signing
+/// methods directly.
+///
+/// Synchronous signers should implement both this trait and [`SignerSync`].
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[auto_impl(&mut, Box)]
+pub trait Signer: Send + Sync {
+    /// Signs the given hash.
+    async fn sign_hash(&self, hash: &B256) -> Result<Signature>;
+
+    /// Signs the hash of the provided message after prefixing it, as specified in [EIP-191].
+    ///
+    /// [EIP-191]: https://eips.ethereum.org/EIPS/eip-191
+    #[inline]
+    async fn sign_message(&self, message: &[u8]) -> Result<Signature> {
+        self.sign_hash(&eip191_hash_message(message)).await
+    }
+
+    /// Signs the transaction.
+    #[cfg(TODO)] // TODO: TypedTransaction
+    #[inline]
+    async fn sign_transaction(&self, message: &TypedTransaction) -> Result<Signature> {
+        self.sign_hash(&message.sighash()).await
+    }
```

Comment:
Confirming we need to derive RLP here, and that's the main blocker? https://github.com/alloy-rs/alloy/blob/main/crates/rpc-types/src/eth/transaction/typed.rs#L18

Reply:
no, a transaction request is not a transaction. transaction requests may contain incomplete data, and are never RLP serialized

Reply:
> builder style Request types (Optional fields) that have a corresponding Transaction type (valid Transaction).

yes, love this plan :)

> going from Request -> Transaction is what we called fill in ethers.

This is a little more complex, I think. The `TransactionRequest` generally allows a `from` field, which the `Transaction` that implements RLP should not

So we may want something like:
- `Transaction` - complete, no from field, implements RLP, no serde
- `SignedTransaction` - complete, no from field, contains Transaction + Signature, no serde
- TransactionRequest - builder for the JSON request object
- `Provider::fill(&mut TransactionRequest)` 
    - populates all possible fields from RPC node
    - should ensure that `TransactionRequest::try_into_transaction` definitely succeeds
- `Signer::sign_transaction(&TransactionRequest) -> Result<SignedTransaction>` 
    - errors when the request is missing info, or if `from ` is populated with address signer can't sign from 
- `TransactionResponse` - JSON object returned by `eth_getTransactionByXXXX` methods 
    - should ALWAYS contain a `SignedTransaction`

Reply:
OK. Supportive of this. Let's do in follow-up @DaniPopes so we can merge this in?